http.get() in init context causes nil pointer dereference panic

Running this script:
```js
import http from 'k6/http';

http.get("http://test.loadimpact.com");

export default function () {
	console.log("test");
}
```
panics with:
```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: Panic at 40: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: Panic at 40: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x120 pc=0xaa5915]
...
```
Not sure if we should allow http function calls in the init context at all, but either we have to show a nice error message or it should work.

--- a/js/common/init_error.go
+++ b/js/common/init_error.go
@@ -1,0 +1,17 @@
+package common
+
+// InitError is an error that happened during the a test initiation
+type InitError string
+
+// NewInitError returns a new InitError with the provided message
+func NewInitError(msg string) InitError {
+	return (InitError)(msg)
+}
+
+func (i InitError) Error() string {
+	return (string)(i)
+}
+
+func (i InitError) String() string {
+	return (string)(i)
+}

--- a/js/common/init_error.go
+++ b/js/common/init_error.go
@@ -1,17 +1,17 @@
 package common
 
-// InitError is an error that happened during the a test initiation
-type InitError string
+// InitContextError is an error that happened during the a test init context
+type InitContextError string
 
-// NewInitError returns a new InitError with the provided message
-func NewInitError(msg string) InitError {
-	return (InitError)(msg)
+// NewInitContextError returns a new InitContextError with the provided message
+func NewInitContextError(msg string) InitContextError {
+	return (InitContextError)(msg)
 }
 
-func (i InitError) Error() string {
+func (i InitContextError) Error() string {
 	return (string)(i)
 }
 
-func (i InitError) String() string {
+func (i InitContextError) String() string {
 	return (string)(i)
 }

--- a/js/modules/k6/http/http.go
+++ b/js/modules/k6/http/http.go
@@ -22,6 +22,7 @@ package http
 
 import (
 	"context"
+	"errors"
 	"net/http"
 	"net/http/cookiejar"
 
@@ -42,6 +43,9 @@ const (
 	HTTP_METHOD_PATCH   = "PATCH"
 	HTTP_METHOD_OPTIONS = "OPTIONS"
 )
+
+// ErrJarForbiddenInInitContext is used when a cookie jar was made in the init context
+var ErrJarForbiddenInInitContext = errors.New("Making cookie jars in the init context is not supported")
 
 type HTTPCookie struct {
 	Name, Value, Domain, Path string
@@ -104,9 +108,12 @@ func (*HTTP) XCookieJar(ctx *context.Context) *HTTPCookieJar {
 	return newCookieJar(ctx)
 }
 
-func (*HTTP) CookieJar(ctx context.Context) *HTTPCookieJar {
+func (*HTTP) CookieJar(ctx context.Context) (*HTTPCookieJar, error) {
 	state := common.GetState(ctx)
-	return &HTTPCookieJar{state.CookieJar, &ctx}
+	if state == nil {
+		return nil, ErrJarForbiddenInInitContext
+	}
+	return &HTTPCookieJar{state.CookieJar, &ctx}, nil
 }
 
 func (*HTTP) mergeCookies(req *http.Request, jar *cookiejar.Jar, reqCookies map[string]*HTTPRequestCookie) map[string][]*HTTPRequestCookie {

--- a/js/modules/k6/http/http.go
+++ b/js/modules/k6/http/http.go
@@ -22,7 +22,6 @@ package http
 
 import (
 	"context"
-	"errors"
 	"net/http"
 	"net/http/cookiejar"
 
@@ -45,7 +44,7 @@ const (
 )
 
 // ErrJarForbiddenInInitContext is used when a cookie jar was made in the init context
-var ErrJarForbiddenInInitContext = errors.New("Making cookie jars in the init context is not supported")
+var ErrJarForbiddenInInitContext = common.NewInitContextError("Making cookie jars in the init context is not supported")
 
 type HTTPCookie struct {
 	Name, Value, Domain, Path string

--- a/js/modules/k6/http/http_request.go
+++ b/js/modules/k6/http/http_request.go
@@ -25,6 +25,7 @@ import (
 	"compress/gzip"
 	"compress/zlib"
 	"context"
+	"errors"
 	"fmt"
 	"io"
 	"io/ioutil"
@@ -49,6 +50,9 @@ import (
 	log "github.com/sirupsen/logrus"
 	null "gopkg.in/guregu/null.v3"
 )
+
+// ErrHTTPForbiddenInInitContext is used when a http requests was made in the init context
+var ErrHTTPForbiddenInInitContext = errors.New("Making http requests in the init context is not supported")
 
 type HTTPRequest struct {
 	Method  string                          `json:"method"`
@@ -156,6 +160,9 @@ type parsedHTTPRequest struct {
 func (h *HTTP) parseRequest(ctx context.Context, method string, reqURL URL, body interface{}, params goja.Value) (*parsedHTTPRequest, error) {
 	rt := common.GetRuntime(ctx)
 	state := common.GetState(ctx)
+	if state == nil {
+		return nil, ErrHTTPForbiddenInInitContext
+	}
 
 	result := &parsedHTTPRequest{
 		url: &reqURL,

--- a/js/modules/k6/http/http_request.go
+++ b/js/modules/k6/http/http_request.go
@@ -54,6 +54,9 @@ import (
 // ErrHTTPForbiddenInInitContext is used when a http requests was made in the init context
 var ErrHTTPForbiddenInInitContext = errors.New("Making http requests in the init context is not supported")
 
+// ErrBatchForbiddenInInitContext is used when batch was made in the init context
+var ErrBatchForbiddenInInitContext = errors.New("Using batch in the init context is not supported")
+
 type HTTPRequest struct {
 	Method  string                          `json:"method"`
 	URL     string                          `json:"url"`
@@ -629,6 +632,9 @@ func (h *HTTP) request(ctx context.Context, preq *parsedHTTPRequest) (*HTTPRespo
 func (h *HTTP) Batch(ctx context.Context, reqsV goja.Value) (goja.Value, error) {
 	rt := common.GetRuntime(ctx)
 	state := common.GetState(ctx)
+	if state == nil {
+		return nil, ErrBatchForbiddenInInitContext
+	}
 
 	// Return values; retval must be guarded by the mutex.
 	var mutex sync.Mutex

--- a/js/modules/k6/http/http_request.go
+++ b/js/modules/k6/http/http_request.go
@@ -25,7 +25,6 @@ import (
 	"compress/gzip"
 	"compress/zlib"
 	"context"
-	"errors"
 	"fmt"
 	"io"
 	"io/ioutil"
@@ -52,10 +51,10 @@ import (
 )
 
 // ErrHTTPForbiddenInInitContext is used when a http requests was made in the init context
-var ErrHTTPForbiddenInInitContext = errors.New("Making http requests in the init context is not supported")
+var ErrHTTPForbiddenInInitContext = common.NewInitError("Making http requests in the init context is not supported")
 
 // ErrBatchForbiddenInInitContext is used when batch was made in the init context
-var ErrBatchForbiddenInInitContext = errors.New("Using batch in the init context is not supported")
+var ErrBatchForbiddenInInitContext = common.NewInitError("Using batch in the init context is not supported")
 
 type HTTPRequest struct {
 	Method  string                          `json:"method"`

--- a/js/modules/k6/http/http_request.go
+++ b/js/modules/k6/http/http_request.go
@@ -51,10 +51,10 @@ import (
 )
 
 // ErrHTTPForbiddenInInitContext is used when a http requests was made in the init context
-var ErrHTTPForbiddenInInitContext = common.NewInitError("Making http requests in the init context is not supported")
+var ErrHTTPForbiddenInInitContext = common.NewInitContextError("Making http requests in the init context is not supported")
 
 // ErrBatchForbiddenInInitContext is used when batch was made in the init context
-var ErrBatchForbiddenInInitContext = common.NewInitError("Using batch in the init context is not supported")
+var ErrBatchForbiddenInInitContext = common.NewInitContextError("Using batch in the init context is not supported")
 
 type HTTPRequest struct {
 	Method  string                          `json:"method"`

--- a/js/modules/k6/k6.go
+++ b/js/modules/k6/k6.go
@@ -36,11 +36,11 @@ import (
 
 type K6 struct{}
 
-// ErrGroupInInitContext is returned when group are using in init context
-var ErrGroupInInitContext = errors.New("Using group in init context is not supported")
+// ErrGroupInInitContext is returned when group() are using in the init context
+var ErrGroupInInitContext = common.NewInitError("Using group() in the init context is not supported")
 
-// ErrCheckInInitContext is returned when check are using in init context
-var ErrCheckInInitContext = errors.New("Using check in init context is not supported")
+// ErrCheckInInitContext is returned when check() are using in the init context
+var ErrCheckInInitContext = common.NewInitError("Using check() in the init context is not supported")
 
 func New() *K6 {
 	return &K6{}

--- a/js/modules/k6/k6.go
+++ b/js/modules/k6/k6.go
@@ -37,10 +37,10 @@ import (
 type K6 struct{}
 
 // ErrGroupInInitContext is returned when group() are using in the init context
-var ErrGroupInInitContext = common.NewInitError("Using group() in the init context is not supported")
+var ErrGroupInInitContext = common.NewInitContextError("Using group() in the init context is not supported")
 
 // ErrCheckInInitContext is returned when check() are using in the init context
-var ErrCheckInInitContext = common.NewInitError("Using check() in the init context is not supported")
+var ErrCheckInInitContext = common.NewInitContextError("Using check() in the init context is not supported")
 
 func New() *K6 {
 	return &K6{}

--- a/js/modules/k6/k6.go
+++ b/js/modules/k6/k6.go
@@ -36,6 +36,12 @@ import (
 
 type K6 struct{}
 
+// ErrGroupInInitContext is returned when group are using in init context
+var ErrGroupInInitContext = errors.New("Using group in init context is not supported")
+
+// ErrCheckInInitContext is returned when check are using in init context
+var ErrCheckInInitContext = errors.New("Using check in init context is not supported")
+
 func New() *K6 {
 	return &K6{}
 }
@@ -62,6 +68,9 @@ func (*K6) RandomSeed(ctx context.Context, seed int64) {
 
 func (*K6) Group(ctx context.Context, name string, fn goja.Callable) (goja.Value, error) {
 	state := common.GetState(ctx)
+	if state == nil {
+		return nil, ErrGroupInInitContext
+	}
 
 	g, err := state.Group.Group(name)
 	if err != nil {
@@ -99,6 +108,9 @@ func (*K6) Group(ctx context.Context, name string, fn goja.Callable) (goja.Value
 
 func (*K6) Check(ctx context.Context, arg0, checks goja.Value, extras ...goja.Value) (bool, error) {
 	state := common.GetState(ctx)
+	if state == nil {
+		return false, ErrCheckInInitContext
+	}
 	rt := common.GetRuntime(ctx)
 	t := time.Now()
 

--- a/js/modules/k6/metrics/metrics.go
+++ b/js/modules/k6/metrics/metrics.go
@@ -35,7 +35,7 @@ type Metric struct {
 }
 
 // ErrMetricsAddInInitContext is error returned when adding to metric is done in the init context
-var ErrMetricsAddInInitContext = errors.New("Adding to metrics in init context is not supported")
+var ErrMetricsAddInInitContext = common.NewInitError("Adding to metrics in the init context is not supported")
 
 func newMetric(ctxPtr *context.Context, name string, t stats.MetricType, isTime []bool) (interface{}, error) {
 	if common.GetState(*ctxPtr) != nil {

--- a/js/modules/k6/metrics/metrics.go
+++ b/js/modules/k6/metrics/metrics.go
@@ -35,7 +35,7 @@ type Metric struct {
 }
 
 // ErrMetricsAddInInitContext is error returned when adding to metric is done in the init context
-var ErrMetricsAddInInitContext = common.NewInitError("Adding to metrics in the init context is not supported")
+var ErrMetricsAddInInitContext = common.NewInitContextError("Adding to metrics in the init context is not supported")
 
 func newMetric(ctxPtr *context.Context, name string, t stats.MetricType, isTime []bool) (interface{}, error) {
 	if common.GetState(*ctxPtr) != nil {

--- a/js/modules/k6/metrics/metrics.go
+++ b/js/modules/k6/metrics/metrics.go
@@ -34,6 +34,9 @@ type Metric struct {
 	metric *stats.Metric
 }
 
+// ErrMetricsAddInInitContext is error returned when adding to metric is done in the init context
+var ErrMetricsAddInInitContext = errors.New("Adding to metrics in init context is not supported")
+
 func newMetric(ctxPtr *context.Context, name string, t stats.MetricType, isTime []bool) (interface{}, error) {
 	if common.GetState(*ctxPtr) != nil {
 		return nil, errors.New("Metrics must be declared in the init context")
@@ -48,8 +51,11 @@ func newMetric(ctxPtr *context.Context, name string, t stats.MetricType, isTime 
 	return common.Bind(rt, Metric{stats.New(name, t, valueType)}, ctxPtr), nil
 }
 
-func (m Metric) Add(ctx context.Context, v goja.Value, addTags ...map[string]string) {
+func (m Metric) Add(ctx context.Context, v goja.Value, addTags ...map[string]string) (bool, error) {
 	state := common.GetState(ctx)
+	if state == nil {
+		return false, ErrMetricsAddInInitContext
+	}
 
 	tags := state.Options.RunTags.CloneTags()
 	if state.Options.SystemTags["group"] {
@@ -68,6 +74,7 @@ func (m Metric) Add(ctx context.Context, v goja.Value, addTags ...map[string]str
 	}
 
 	stats.PushIfNotCancelled(ctx, state.Samples, stats.Sample{Time: time.Now(), Metric: m.metric, Value: vfloat, Tags: stats.IntoSampleTags(&tags)})
+	return true, nil
 }
 
 type Metrics struct{}

--- a/js/modules/k6/ws/ws.go
+++ b/js/modules/k6/ws/ws.go
@@ -39,8 +39,8 @@ import (
 	"github.com/loadimpact/k6/stats"
 )
 
-// ErrWSInInitContext is returned when websockets are using in init context
-var ErrWSInInitContext = errors.New("Using websockets in init context is not supported")
+// ErrWSInInitContext is returned when websockets are using in the init context
+var ErrWSInInitContext = common.NewInitError("Using websockets in the init context is not supported")
 
 type WS struct{}
 

--- a/js/modules/k6/ws/ws.go
+++ b/js/modules/k6/ws/ws.go
@@ -40,7 +40,7 @@ import (
 )
 
 // ErrWSInInitContext is returned when websockets are using in the init context
-var ErrWSInInitContext = common.NewInitError("Using websockets in the init context is not supported")
+var ErrWSInInitContext = common.NewInitContextError("Using websockets in the init context is not supported")
 
 type WS struct{}
 

--- a/js/modules/k6/ws/ws.go
+++ b/js/modules/k6/ws/ws.go
@@ -39,6 +39,9 @@ import (
 	"github.com/loadimpact/k6/stats"
 )
 
+// ErrWSInInitContext is returned when websockets are using in init context
+var ErrWSInInitContext = errors.New("Using websockets in init context is not supported")
+
 type WS struct{}
 
 type pingDelta struct {
@@ -79,6 +82,9 @@ func New() *WS {
 func (*WS) Connect(ctx context.Context, url string, args ...goja.Value) (*WSHTTPResponse, error) {
 	rt := common.GetRuntime(ctx)
 	state := common.GetState(ctx)
+	if state == nil {
+		return nil, ErrWSInInitContext
+	}
 
 	// The params argument is optional
 	var callableV, paramsV goja.Value

--- a/js/runner_test.go
+++ b/js/runner_test.go
@@ -35,6 +35,7 @@ import (
 	"github.com/loadimpact/k6/core"
 	"github.com/loadimpact/k6/core/local"
 	"github.com/loadimpact/k6/js/common"
+	k6http "github.com/loadimpact/k6/js/modules/k6/http"
 	"github.com/loadimpact/k6/lib"
 	"github.com/loadimpact/k6/lib/metrics"
 	"github.com/loadimpact/k6/lib/testutils"
@@ -1217,4 +1218,27 @@ func TestVUIntegrationClientCerts(t *testing.T) {
 			})
 		}
 	})
+}
+
+func TestHTTPRequestInInitContext(t *testing.T) {
+	tb := testutils.NewHTTPMultiBin(t)
+	defer tb.Cleanup()
+
+	_, err := New(&lib.SourceData{
+		Filename: "/script.js",
+		Data: []byte(tb.Replacer.Replace(`
+					import { check, fail } from "k6";
+					import http from "k6/http";
+					let res = http.get("HTTPBIN_URL/");
+					export default function() {
+						console.log(test);ci
+					}
+				`)),
+	}, afero.NewMemMapFs(), lib.RuntimeOptions{})
+	if assert.Error(t, err) {
+		assert.Equal(
+			t,
+			"GoError: "+k6http.ErrHTTPForbiddenInInitContext.Error(),
+			err.Error())
+	}
 }

--- a/js/runner_test.go
+++ b/js/runner_test.go
@@ -35,7 +35,10 @@ import (
 	"github.com/loadimpact/k6/core"
 	"github.com/loadimpact/k6/core/local"
 	"github.com/loadimpact/k6/js/common"
+	"github.com/loadimpact/k6/js/modules/k6"
 	k6http "github.com/loadimpact/k6/js/modules/k6/http"
+	k6metrics "github.com/loadimpact/k6/js/modules/k6/metrics"
+	"github.com/loadimpact/k6/js/modules/k6/ws"
 	"github.com/loadimpact/k6/lib"
 	"github.com/loadimpact/k6/lib/metrics"
 	"github.com/loadimpact/k6/lib/testutils"
@@ -1240,5 +1243,84 @@ func TestHTTPRequestInInitContext(t *testing.T) {
 			t,
 			"GoError: "+k6http.ErrHTTPForbiddenInInitContext.Error(),
 			err.Error())
+	}
+}
+
+func TestInitContextForbidden(t *testing.T) {
+	var table = [...][3]string{
+		{
+			"http.request",
+			`import http from "k6/http";
+			 let res = http.get("HTTPBIN_URL");
+			 export default function() { console.log("p"); }`,
+			k6http.ErrHTTPForbiddenInInitContext.Error(),
+		},
+		{
+			"http.batch",
+			`import http from "k6/http";
+			 let res = http.batch("HTTPBIN_URL/something", "HTTPBIN_URL/else");
+			 export default function() { console.log("p"); }`,
+			k6http.ErrBatchForbiddenInInitContext.Error(),
+		},
+		{
+			"http.cookieJar",
+			`import http from "k6/http";
+			 let jar = http.cookieJar();
+			 export default function() { console.log("p"); }`,
+			k6http.ErrJarForbiddenInInitContext.Error(),
+		},
+		{
+			"check",
+			`import { check } from "k6";
+			 check("test", {'is test': (test) => test == "test"})
+			 export default function() { console.log("p"); }`,
+			k6.ErrCheckInInitContext.Error(),
+		},
+		{
+			"group",
+			`import { group } from "k6";
+			 group("group1", function () { console.log("group1");})
+			 export default function() { console.log("p"); }`,
+			k6.ErrGroupInInitContext.Error(),
+		},
+		{
+			"ws",
+			`import ws from "k6/ws";
+			 var url = "ws://echo.websocket.org";
+			 var params = { "tags": { "my_tag": "hello" } };
+			 var response = ws.connect(url, params, function (socket) {
+			   socket.on('open', function open() {
+					console.log('connected');
+			   })
+		   });
+
+			 export default function() { console.log("p"); }`,
+			ws.ErrWSInInitContext.Error(),
+		},
+		{
+			"metric",
+			`import { Counter } from "k6/metrics";
+			 let counter = Counter("myCounter");
+			 counter.add(1);
+			 export default function() { console.log("p"); }`,
+			k6metrics.ErrMetricsAddInInitContext.Error(),
+		},
+	}
+	tb := testutils.NewHTTPMultiBin(t)
+	defer tb.Cleanup()
+
+	for _, test := range table {
+		t.Run(test[0], func(t *testing.T) {
+			_, err := New(&lib.SourceData{
+				Filename: "/script.js",
+				Data:     []byte(tb.Replacer.Replace(test[1])),
+			}, afero.NewMemMapFs(), lib.RuntimeOptions{})
+			if assert.Error(t, err) {
+				assert.Equal(
+					t,
+					"GoError: "+test[2],
+					err.Error())
+			}
+		})
 	}
 }

--- a/release notes/upcoming.md
+++ b/release notes/upcoming.md
@@ -116,3 +116,6 @@ Thanks to @AndriiChuzhynov for implementing this! (#766)
 * Setup data: Distinguish between `undefined` (when there is no `setup()` function or when it doesn't return anything) and `null` (when `setup()` explicitly returns `null`) values for the setup `data` that's passed to the default function and `teardown()` (#799)
 * Setup data: Prevent data races by having each VU have its own independent copy of the setup data (#799)
 * HAR converter: Support HAR files that don't have a `pages` array (#806)
+* UX: Instead of panicing on some operations in the init context we not return an error that the given
+  action is not supported(#811). This includes: making http requests(batched or not), websockets,
+  adding to metrics, making checks and groups or initializing cookie jars.

